Gemfile

```diff
@@ -6,9 +6,9 @@ ruby "2.6.3"
 # Bundle edge Rails instead: gem "rails", github: "rails/rails"
 gem "figaro"
 
-gem "rails", "~> 6.0.0"
-
+gem "bcrypt", "~> 3.1.12"
```

Comment:
gem này dùng làm gì em

Reply:
gem "figaro" : dùng để quản lý các biến môi trường, lưu trữ các biến local không cho người ngoài biết.
gem "bcrypt": dùng để mã hóa mật khẩu. Đây là thuật toán mã hóa 1 chiều giúp bảo mật password, gồm 3 phần:
+ Password người dùng
+ Cost mặc định là 10. Chuỗi càng dài thì hash càng khó tìm.
+ Chuỗi salt random bởi bcrypt.
Sau đó Hash password + salt + cost và lưu vào database. Chuỗi hash 22 kí tự đầu là salt, tiếp đến là số kí tự cost rồi đến password của người dùng. Còn khi kiểm tra thì bcrypt nhờ vào 3 phần trên mà hash ra password để so sánh.